File: native/android/PropertyCross/src/com/propertycross/android/services/GeoLocationService.java
package com.propertycross.android.services;

import java.util.Date;

import android.location.Criteria;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;

import com.propertycross.android.events.Callback;
import com.propertycross.android.presenter.GeoLocation;
import com.propertycross.android.presenter.IGeoLocationService;
import com.propertycross.android.presenter.IMarshalInvokeService;

public class GeoLocationService implements IGeoLocationService, LocationListener {

	private final long FIVE_MINUTES = 1000 * 60 * 5;
	private final long SEVEN_SECONDS = 1000 * 7;
	private LocationManager manager;
	private Callback<GeoLocation> callback;
	private IMarshalInvokeService marshal;
	
	public GeoLocationService(LocationManager locationManager, IMarshalInvokeService marshalService) {
		this.manager = locationManager;
		this.marshal = marshalService;
	}
	
	public void getLocation(Callback<GeoLocation> callback) {
		this.callback = callback;
		
		Criteria c = new Criteria();
		c.setAccuracy(Criteria.ACCURACY_FINE);
		
		String provider = manager.getBestProvider(c, true);
		if(provider != null) {
			doCallback(null);
		}
		else {
			Location l = manager.getLastKnownLocation(provider);
			if (isLastLocationAccurateEnough(l)) {
				doCallback(toGeoLocation(l));
			}
			
			manager.requestLocationUpdates(provider, 1000, 10, this);
		}
	}
	
	private boolean isLastLocationAccurateEnough(Location l) {
		return l != null && l.getTime() >= unixTimeNow() - FIVE_MINUTES;
	}
	
	private long unixTimeNow() {
		return new Date().getTime();
	}
	
	private GeoLocation toGeoLocation(Location l) {
		return new GeoLocation(l.getLatitude(), l.getLongitude());
	}
	
	private void doCallback(GeoLocation g) {
		if (callback != null) {
			callback.complete(g);
			callback = null;
		}
	}	
	
	@Override
	public void onLocationChanged(Location location) {
		unsubscribe();
		
		if (location != null) {
			doCallback(toGeoLocation(location));
		}		
	}

	@Override
	public void onProviderDisabled(String provider) {
		unsubscribe();
	}

	@Override
	public void onProviderEnabled(String provider) {
	}

	@Override
	public void onStatusChanged(String provider, int status, Bundle extras) {
	}
	
	public void unsubscribe() {
		if (manager != null) {
			manager.removeUpdates(this);
		}
	}
}

File: xamarin-Mvvmcross/PropertyCross.Droid/Services/GeoLocationService.cs
using System;
using System.Timers;
using Android.Locations;
using Android.OS;
using PropertyCross.Core.Domain.Services;

namespace PropertyCross.Droid.Services
{
	public class GeoLocationService : Java.Lang.Object, ILocationListener, IGeoLocationService, IDisposable
	{
		private static readonly long FIVE_MINUTES = 1000 * 60 * 5;
		private static readonly long SEVEN_SECONDS = 1000 * 7;
		private static readonly DateTime Epoch = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
		private LocationManager manager;
		private Action<GeoLocation> pendingCallback;
		private Timer timer;
		private IMarshalInvokeService marshal;

		public GeoLocationService(LocationManager locationManager, IMarshalInvokeService marshalService)
		{
			manager = locationManager;
			marshal = marshalService;
			timer = new Timer(SEVEN_SECONDS);
			timer.Elapsed += OnTimeout;
		}

		public void GetLocation(Action<GeoLocation> callback)
		{
			pendingCallback = callback;

			Criteria c = new Criteria()
			{
				Accuracy = Accuracy.Fine,
			};

			string provider = manager.GetBestProvider(c, true);
			if(string.IsNullOrEmpty(provider))
			{
				DoCallback(null);
			}
			else
			{
				Location l = manager.GetLastKnownLocation(provider);
				if(IsLastLocationAccurateEnough(l))
				{
					DoCallback(ToGeolocation(l));
				}

				manager.RequestLocationUpdates(provider, 1000, 10, this);
				timer.Start();
			}
		}

		private bool IsLastLocationAccurateEnough(Location l)
		{
			return l != null &&
				l.Time >= UnixTimeNow() - FIVE_MINUTES;
		}

		private long UnixTimeNow()
		{
			return (DateTime.Now.Ticks - Epoch.Ticks) / TimeSpan.TicksPerMillisecond;
		}

		private GeoLocation ToGeolocation(Location location)
		{
			return new GeoLocation ()
			{
				Latitude = location.Latitude,
				Longitude = location.Longitude
			};
		}

		private void OnTimeout(object sender, ElapsedEventArgs e)
		{
			marshal.Invoke(() =>
			{
				Unsubscribe();
				DoCallback(null);
			});
		}

		public void OnLocationChanged(Location location)
		{
			Unsubscribe();

			if (location != null)
			{
				DoCallback(ToGeolocation(location));
			}
		}
		
		public void OnProviderDisabled(string provider)
		{
			Unsubscribe();
		}
		
		public void OnProviderEnabled(string provider)
		{
		}
		
		public void OnStatusChanged(string provider, Availability status, Bundle extras)
		{
		}

		public void Unsubscribe()
		{
			if (manager != null)
				manager.RemoveUpdates(this);
			timer.Stop();
		}

		private void DoCallback(GeoLocation g)
		{
			if(pendingCallback != null)
			{
				pendingCallback(g);
				pendingCallback = null;
			}
		}

		public void Dispose()
		{
			base.Dispose();
			timer.Elapsed -= OnTimeout;
			timer.Dispose();
		}
	}
}